hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/AbstractRpcClient.java

```diff
@@ -491,13 +482,13 @@ public void close() {
       connToClose = connections.values();
       connections.clear();
     }
-    cleanupIdleConnectionTask.cancel(true);
     for (T conn : connToClose) {
-      conn.shutdown();
+      // conn may be null in case of cancellation
+      if (conn != null) conn.shutdown();
```

Comment:
I do not fully understand, here the connection is gotten from a Map, we will not put null element to the Map?

Reply:
The key in the map is not null, but the value ....

`connections.removeValue(conn.remoteId(), conn);`

I tried to contribute this patch because i get around a 10x throughput and a lot of error/warn message less.
By "unlocking" the client bottleneck, i found a server bug (HBASE-22492) and solve it.

Now, if to share this i have to spend a lot of time to justify such peace of code, i will give up.
For my own use, this patch is already in production.

Reply:
That's fine, you can just share your ideas on how to get the 10x throughput. But for a open source project we have to make sure that all the parts work correctly, as you may only use A so you do not care B but some others may use B...

And on the removal, it will completely remove the entry from the map, instead of setting the value to null I believe?

Reply:
removeValue just removes the value (set it as null)

org/apache/hadoop/hbase/util/PoolMap.java

```
  public boolean removeValue(K key, V value) {
    Pool<V> pool = pools.get(key);
    boolean res = false;
    if (pool != null) {
      res = pool.remove(value);
      if (res && pool.size() == 0) {
        pools.remove(key);
      }
    }
    return res;
  }
```

Just one precision: i found it, and added this if because at runtime i get a NPE. Not just for fun ...

Reply:
The code you paste here do not set the value to null I think? If you actually hit an NPE, then I think there must be something wrong at other places. Looking at the code for PoolMap, which pool type do you use?

Reply:
Indeed, and i didn't succeed to reproduce the NPE which was probably a consequence of all my different try when fixing HBASE-22492 ....

Reply:
Corrected